File: python/src/falcon_core/math/arrays/measured_array.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.measured_array import MeasuredArray as _CMeasuredArray
from falcon_core.generic.f_array import FArray
from falcon_core.generic.list import List

class MeasuredArray:
    """Python wrapper for MeasuredArray."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def from_json(cls, json: str) -> MeasuredArray:
        return cls(_CMeasuredArray.from_json(json))

    @classmethod
    def from_data(cls, data: Any, shape: Any, ndim: Any) -> MeasuredArray:
        ret = _CMeasuredArray.from_data(data, shape, ndim)
        return MeasuredArray._from_capi(ret)

    @classmethod
    def from_farray(cls, farray: FArray) -> MeasuredArray:
        ret = _CMeasuredArray.from_farray(farray._c if farray is not None else None)
        return MeasuredArray._from_capi(ret)

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def dimension(self, ) -> None:
        ret = self._c.dimension()
        return ret

    def shape(self, out_buffer: Any, ndim: Any) -> None:
        ret = self._c.shape(out_buffer, ndim)
        return ret

    def data(self, out_buffer: Any, numdata: Any) -> None:
        ret = self._c.data(out_buffer, numdata)
        return ret

    def plus_equals_farray(self, other: FArray) -> None:
        ret = self._c.plus_equals_farray(other._c if other is not None else None)
        return ret

    def plus_equals_double(self, other: Any) -> None:
        ret = self._c.plus_equals_double(other)
        return ret

    def plus_equals_int(self, other: Any) -> None:
        ret = self._c.plus_equals_int(other)
        return ret

    def plus_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.plus_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def plus_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.plus_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def plus_double(self, other: Any) -> MeasuredArray:
        ret = self._c.plus_double(other)
        return MeasuredArray._from_capi(ret)

    def plus_int(self, other: Any) -> MeasuredArray:
        ret = self._c.plus_int(other)
        return MeasuredArray._from_capi(ret)

    def minus_equals_farray(self, other: FArray) -> None:
        ret = self._c.minus_equals_farray(other._c if other is not None else None)
        return ret

    def minus_equals_double(self, other: Any) -> None:
        ret = self._c.minus_equals_double(other)
        return ret

    def minus_equals_int(self, other: Any) -> None:
        ret = self._c.minus_equals_int(other)
        return ret

    def minus_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.minus_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def minus_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.minus_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def minus_double(self, other: Any) -> MeasuredArray:
        ret = self._c.minus_double(other)
        return MeasuredArray._from_capi(ret)

    def minus_int(self, other: Any) -> MeasuredArray:
        ret = self._c.minus_int(other)
        return MeasuredArray._from_capi(ret)

    def negation(self, ) -> MeasuredArray:
        ret = self._c.negation()
        return MeasuredArray._from_capi(ret)

    def times_equals_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.times_equals_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def times_equals_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.times_equals_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def times_equals_double(self, other: Any) -> None:
        ret = self._c.times_equals_double(other)
        return ret

    def times_equals_int(self, other: Any) -> None:
        ret = self._c.times_equals_int(other)
        return ret

    def times_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.times_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def times_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.times_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def times_double(self, other: Any) -> MeasuredArray:
        ret = self._c.times_double(other)
        return MeasuredArray._from_capi(ret)

    def times_int(self, other: Any) -> MeasuredArray:
        ret = self._c.times_int(other)
        return MeasuredArray._from_capi(ret)

    def divides_equals_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.divides_equals_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def divides_equals_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.divides_equals_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def divides_equals_double(self, other: Any) -> None:
        ret = self._c.divides_equals_double(other)
        return ret

    def divides_equals_int(self, other: Any) -> None:
        ret = self._c.divides_equals_int(other)
        return ret

    def divides_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.divides_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def divides_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.divides_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def divides_double(self, other: Any) -> MeasuredArray:
        ret = self._c.divides_double(other)
        return MeasuredArray._from_capi(ret)

    def divides_int(self, other: Any) -> MeasuredArray:
        ret = self._c.divides_int(other)
        return MeasuredArray._from_capi(ret)

    def pow(self, other: Any) -> MeasuredArray:
        ret = self._c.pow(other)
        return MeasuredArray._from_capi(ret)

    def abs(self, ) -> MeasuredArray:
        ret = self._c.abs()
        return MeasuredArray._from_capi(ret)

    def min(self, ) -> None:
        ret = self._c.min()
        return ret

    def min_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.min_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def min_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.min_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def max(self, ) -> None:
        ret = self._c.max()
        return ret

    def max_farray(self, other: FArray) -> MeasuredArray:
        ret = self._c.max_farray(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def max_measured_array(self, other: MeasuredArray) -> MeasuredArray:
        ret = self._c.max_measured_array(other._c if other is not None else None)
        return MeasuredArray._from_capi(ret)

    def equal(self, other: MeasuredArray) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: MeasuredArray) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def greater_than(self, value: Any) -> None:
        ret = self._c.greater_than(value)
        return ret

    def less_than(self, value: Any) -> None:
        ret = self._c.less_than(value)
        return ret

    def remove_offset(self, offset: Any) -> None:
        ret = self._c.remove_offset(offset)
        return ret

    def sum(self, ) -> None:
        ret = self._c.sum()
        return ret

    def reshape(self, shape: Any, ndims: Any) -> MeasuredArray:
        ret = self._c.reshape(shape, ndims)
        return MeasuredArray._from_capi(ret)

    def where(self, value: Any) -> List:
        ret = self._c.where(value)
        if ret is None: return None
        return List(ret)

    def flip(self, axis: Any) -> MeasuredArray:
        ret = self._c.flip(axis)
        return MeasuredArray._from_capi(ret)

    def full_gradient(self, out_buffer: MeasuredArray, buffer_size: Any) -> None:
        ret = self._c.full_gradient(out_buffer._c if out_buffer is not None else None, buffer_size)
        return ret

    def gradient(self, axis: Any) -> MeasuredArray:
        ret = self._c.gradient(axis)
        return MeasuredArray._from_capi(ret)

    def get_sum_of_squares(self, ) -> None:
        ret = self._c.get_sum_of_squares()
        return ret

    def get_summed_diff_int_of_squares(self, other: Any) -> None:
        ret = self._c.get_summed_diff_int_of_squares(other)
        return ret

    def get_summed_diff_double_of_squares(self, other: Any) -> None:
        ret = self._c.get_summed_diff_double_of_squares(other)
        return ret

    def get_summed_diff_array_of_squares(self, other: MeasuredArray) -> None:
        ret = self._c.get_summed_diff_array_of_squares(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __len__(self):
        return self.size()

    def __add__(self, other):
        """Operator overload for +"""
        if hasattr(other, "_c") and type(other).__name__ == "EqualsFarray":
            return self.plus_equals_farray(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsDouble":
            return self.plus_equals_double(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsInt":
            return self.plus_equals_int(other)
        if isinstance(other, MeasuredArray):
            return self.plus_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ in ["FArrayDouble", "FArrayInt", "FArray"]:
            return self.plus_farray(other)
        if isinstance(other, float):
            return self.plus_double(other)
        if isinstance(other, int):
            return self.plus_int(other)
        return NotImplemented

    def __sub__(self, other):
        """Operator overload for -"""
        if hasattr(other, "_c") and type(other).__name__ == "EqualsFarray":
            return self.minus_equals_farray(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsDouble":
            return self.minus_equals_double(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsInt":
            return self.minus_equals_int(other)
        if isinstance(other, MeasuredArray):
            return self.minus_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ in ["FArrayDouble", "FArrayInt", "FArray"]:
            return self.minus_farray(other)
        if isinstance(other, float):
            return self.minus_double(other)
        if isinstance(other, int):
            return self.minus_int(other)
        return NotImplemented

    def __mul__(self, other):
        """Operator overload for *"""
        if isinstance(other, MeasuredArray):
            return self.times_equals_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsFarray":
            return self.times_equals_farray(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsDouble":
            return self.times_equals_double(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsInt":
            return self.times_equals_int(other)
        if isinstance(other, MeasuredArray):
            return self.times_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ in ["FArrayDouble", "FArrayInt", "FArray"]:
            return self.times_farray(other)
        if isinstance(other, float):
            return self.times_double(other)
        if isinstance(other, int):
            return self.times_int(other)
        return NotImplemented

    def __truediv__(self, other):
        """Operator overload for /"""
        if isinstance(other, MeasuredArray):
            return self.divides_equals_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsFarray":
            return self.divides_equals_farray(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsDouble":
            return self.divides_equals_double(other)
        if hasattr(other, "_c") and type(other).__name__ == "EqualsInt":
            return self.divides_equals_int(other)
        if isinstance(other, MeasuredArray):
            return self.divides_measured_array(other)
        if hasattr(other, "_c") and type(other).__name__ in ["FArrayDouble", "FArrayInt", "FArray"]:
            return self.divides_farray(other)
        if isinstance(other, float):
            return self.divides_double(other)
        if isinstance(other, int):
            return self.divides_int(other)
        return NotImplemented

    def __neg__(self):
        """Operator overload for unary -"""
        return self.negation()

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, MeasuredArray):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, MeasuredArray):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/src/math/arrays/measuredarray.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class c_measuredarray (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.measuredarray_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module MeasuredArray = struct
  type t = c_measuredarray

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.measuredarray_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.measuredarray_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_measuredarray ptr

  let fromData (data : float) (shape : int) (ndim : int) : t =
    let ptr = Capi_bindings.measuredarray_from_data data (Unsigned.Size_t.of_int shape) (Unsigned.Size_t.of_int ndim) in
    ErrorHandling.raise_if_error ();
    new c_measuredarray ptr

  let fromFarray (farray : Farraydouble.t) : t =
    ErrorHandling.read farray (fun () ->
      let ptr = Capi_bindings.measuredarray_from_farray farray#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let size (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_size handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let dimension (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_dimension handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let shape (handle : t) (out_buffer : int) (ndim : int) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_shape handle#raw (Unsigned.Size_t.of_int out_buffer) (Unsigned.Size_t.of_int ndim) in
      ErrorHandling.raise_if_error ();
      result
    )

  let data (handle : t) (out_buffer : float) (numdata : int) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_data handle#raw out_buffer (Unsigned.Size_t.of_int numdata) in
      ErrorHandling.raise_if_error ();
      result
    )

  let plusEqualsFarray (handle : t) (other : Farraydouble.t) : unit =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_plus_equals_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let plusEqualsDouble (handle : t) (other : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_plus_equals_double handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let plusEqualsInt (handle : t) (other : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_plus_equals_int handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let plusMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_plus_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let plusFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_plus_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let plusDouble (handle : t) (other : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_plus_double handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let plusInt (handle : t) (other : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_plus_int handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let minusEqualsFarray (handle : t) (other : Farraydouble.t) : unit =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_minus_equals_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let minusEqualsDouble (handle : t) (other : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_minus_equals_double handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let minusEqualsInt (handle : t) (other : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_minus_equals_int handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let minusMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_minus_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let minusFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_minus_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let minusDouble (handle : t) (other : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_minus_double handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let minusInt (handle : t) (other : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_minus_int handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let negation (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_negation handle#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesEqualsMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_times_equals_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesEqualsFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_times_equals_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesEqualsDouble (handle : t) (other : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_times_equals_double handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let timesEqualsInt (handle : t) (other : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_times_equals_int handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let timesMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_times_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_times_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesDouble (handle : t) (other : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_times_double handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let timesInt (handle : t) (other : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_times_int handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesEqualsMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_divides_equals_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesEqualsFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_divides_equals_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesEqualsDouble (handle : t) (other : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_divides_equals_double handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let dividesEqualsInt (handle : t) (other : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_divides_equals_int handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let dividesMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_divides_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_divides_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesDouble (handle : t) (other : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_divides_double handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let dividesInt (handle : t) (other : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_divides_int handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let pow (handle : t) (other : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_pow handle#raw other in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let abs (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_abs handle#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let min (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_min handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let minFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_min_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let minMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_min_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let max (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_max handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let maxFarray (handle : t) (other : Farraydouble.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_max_farray handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let maxMeasuredArray (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_max_measured_array handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let greaterThan (handle : t) (value : float) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_greater_than handle#raw value in
      ErrorHandling.raise_if_error ();
      result
    )

  let lessThan (handle : t) (value : float) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_less_than handle#raw value in
      ErrorHandling.raise_if_error ();
      result
    )

  let removeOffset (handle : t) (offset : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_remove_offset handle#raw offset in
      ErrorHandling.raise_if_error ();
      result
    )

  let sum (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_sum handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let reshape (handle : t) (shape : int) (ndims : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_reshape handle#raw (Unsigned.Size_t.of_int shape) (Unsigned.Size_t.of_int ndims) in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let where (handle : t) (value : float) : Listlistsizet.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_where handle#raw value in
      ErrorHandling.raise_if_error ();
      new c_listlistsizet result
    )

  let flip (handle : t) (axis : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_flip handle#raw (Unsigned.Size_t.of_int axis) in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let fullGradient (handle : t) (out_buffer : t) (buffer_size : int) : int =
    ErrorHandling.multi_read [handle; out_buffer] (fun () ->
      let result = Capi_bindings.measuredarray_full_gradient handle#raw out_buffer#raw (Unsigned.Size_t.of_int buffer_size) in
      ErrorHandling.raise_if_error ();
      result
    )

  let gradient (handle : t) (axis : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_gradient handle#raw (Unsigned.Size_t.of_int axis) in
      ErrorHandling.raise_if_error ();
      new c_measuredarray result
    )

  let getSumOfSquares (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_get_sum_of_squares handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let getSummedDiffIntOfSquares (handle : t) (other : int) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_get_summed_diff_int_of_squares handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let getSummedDiffDoubleOfSquares (handle : t) (other : float) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measuredarray_get_summed_diff_double_of_squares handle#raw other in
      ErrorHandling.raise_if_error ();
      result
    )

  let getSummedDiffArrayOfSquares (handle : t) (other : t) : float =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measuredarray_get_summed_diff_array_of_squares handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end